internal/filters.go

```diff
@@ -0,0 +1,208 @@
+func (mf *FilterFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	// passes if event does not match
+	events := make([]*Event, 0)
+	if !mf.Matcher.Match(ctx, event, mf.Pattern) {
+		events = append(events, event)
+	}
+	return events, nil
+}
+
+// Filter transforms an event according to its transformation spec
+func (mf *TransformFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	// identity transform
+	events := make([]*Event, 0)
+	//TODO: implement filter logic
+	events = append(events, event)
+	return events, nil
+}
+
+// Filter filters event if it hhas expired
+func (mf *TTLFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	// never filters
+	events := make([]*Event, 0)
+	//TODO: implement filter logic
+	events = append(events, event)
+	return events, nil
+}
+
+// Filter splits an event containing an array into multiple events
+func (mf *SplitFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	// always splits into two identical events
+	events := make([]*Event, 0)
+	//TODO: implement filter logic
+	//TODO: clone
+	events = append(events, event)
+	events = append(events, event)
+	return events, nil
+}
+
+// Filter lets any event pass
+func (mf *PassFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	return []*Event{event}, nil
+}
+
+// Filter lets no event pass
+func (mf *BlockFilter) Filter(ctx context.Context, event *Event) ([]*Event, error) {
+	return []*Event{}, nil
+}
+
+// NewFilterer is a factory function to create appropriate filterer for given filter plugin config
+func NewFilterer(ctx context.Context, fp *FilterPlugin) (Filterer, error) {
+	if fp == nil {
+		return nil, new(MissingFilterPluginConfigError)
+	}
+	switch fp.Type {
+	case FilterTypeFilter:
+		flt := new(FilterFilter)
+		flt.Matcher = NewDefaultPatternMatcher()
+		// parse config params
+		if fp.Params != nil {
+			if value, ok := fp.Params["pattern"]; ok {
+				flt.Pattern = value
+			}
+		}
+		return flt, nil
+	case FilterTypeMatcher:
+		flt := new(MatchFilter)
+		flt.Matcher = NewDefaultPatternMatcher()
+		// parse config params
+		if fp.Params != nil {
+			if value, ok := fp.Params["pattern"]; ok {
+				flt.Pattern = value
+			}
+		}
+		return flt, nil
+	case FilterTypeTransformer:
+		//TODO: pass in config params here
+		return new(TransformFilter), nil
+	case FilterTypeTTLer:
+		//TODO: pass in config params here
+		return new(TTLFilter), nil
+	case FilterTypeSplitter:
+		//TODO: pass in config params here
+		return new(SplitFilter), nil
+	case FilterTypePass:
+		return new(PassFilter), nil
+	case FilterTypeBlock:
+		return new(BlockFilter), nil
+	}
+	return nil, &UnknownFilterTypeError{fp.Type}
+}
+
+// DoSync synchronoulsy accepts an event into a filter plugin belonging to a filter chain
+func (fp *FilterPlugin) DoSync(ctx context.Context, event *Event) error {
+	log.Debug().Msg(fp.Type + " filter " + fp.Hash(ctx) + " passed")
```

Comment:
We should log with log.Ctx(ctx).Debug() so that context log values are part of the logs

Reply:
ok